src/XR/features/WebXRLayers.ts

```diff
@@ -0,0 +1,222 @@
+        const lastLeftSubImage = this._lastSubImages.get('left');
+        if (!!lastLeftSubImage) {
+            return {
+                framebufferWidth: lastLeftSubImage.textureWidth,
+                framebufferHeight: lastLeftSubImage.textureHeight
+            };
+        }
+        return null;
+    }
+}
+
+/**
+ * Provides render target textures and other important rendering information for a given XRProjectionLayer.
+ * @hidden
+ */
+class XRProjectionLayerRenderTargetProvider extends XRCompositionLayerRenderTargetProvider {
+    constructor(
+        _xrSessionManager: WebXRSessionManager,
+        _xrWebGLBinding: XRWebGLBinding,
+        private readonly _projectionLayer: XRProjectionLayer) {
+        super(_xrSessionManager, _xrWebGLBinding, _projectionLayer);
+    }
+
+    public getRenderTargetForView(view: XRView): Nullable<RenderTargetTexture> {
+        const subImage = this._xrWebGLBinding.getViewSubImage(this._projectionLayer, view);
+        return this._getRenderTargetForSubImage(subImage, view.eye);
+    }
+
+    public getRenderTargetForEye(eye: XREye): Nullable<RenderTargetTexture> {
+        const lastSubImage = this._lastSubImages.get(eye);
+        if (lastSubImage) {
+            return this._getRenderTargetForSubImage(lastSubImage, eye);
+        }
+        return null;
+    }
+
+    public getFramebufferDimensions(): Nullable<{ framebufferWidth: number; framebufferHeight: number; }> {
+        return {
+            framebufferWidth: this._projectionLayer.textureWidth,
+            framebufferHeight: this._projectionLayer.textureHeight
+        };
+    }
+}
+
+/**
+ * Wraps the webxr composition layers interface.
+ */
+export class WebXRLayers extends WebXRAbstractFeature implements WebXRLayerWrapperProvider {
+    /**
+     * The module's name
+     */
+    public static readonly Name = WebXRFeatureName.LAYERS;
+    /**
+     * The (Babylon) version of this module.
+     * This is an integer representing the implementation version.
+     * This number does not correspond to the WebXR specs version
+     */
+    public static readonly Version = 1;
+
+    private _xrWebGLBinding: XRWebGLBinding;
+    private _layerTypes = new Map<any, WebXRCompositionLayerType>();
+
+    constructor(_xrSessionManager: WebXRSessionManager) {
+        super(_xrSessionManager);
+        this.xrNativeFeatureName = "layers";
+
+        _xrSessionManager.onXRSessionInit.add(() => {
+            const glContext = _xrSessionManager.scene.getEngine()._gl;
+            this._xrWebGLBinding = new XRWebGLBinding(_xrSessionManager.session, glContext);
+        });
+
+        // Inject this instance for the session manager to use for wrapping layers.
+        _xrSessionManager._addLayerWrapperProvider(this);
+    }
+
+    /**
+     * Creates a new XRProjectionLayer.
+     * @returns the projection layer
+     */
+    public createProjectionLayer(): XRProjectionLayer {
```

Comment:
Yeah so the only reason you can generate unwrapped layers is so that you can pass them to the session manager `updateRenderState` in an `XRRenderStateInit` object, which only has an array of type `XRLayer` for the layers. Creating the layer doesn't add it to the layers array, you need to add it yourself to an array that you pass to `updateRenderState`.

We currently only support projection layers, so this is a strange example, but you'd do something like:
```typescript
const projectionLayer = layersFeature.createProjectionLayer();
const projectionLayer2 = layersFeature.createProjectionLayer();
// Use one layer
baseExperience.sessionManager.updateRenderState({ layers: [projectionLayer] });
// Use two layers
baseExperience.sessionManager.updateRenderState({ layers: [projectionLayer, projectionLayer2] });
// Remove one layer, use only one layer
baseExperience.sessionManager.updateRenderState({ layers: [projectionLayer] });
```
The current webXRSessionManager functions for retrieving render targets only work for a single layer, which right now is either the `baseLayer` or the first in the `layers` array, but after adding the layer to the session's render state you could in theory generate a render target provider from any of the layers and fetch your render targets directly from there, something like:
```typescript
baseExperience.sessionManager.updateRenderState({ layers: [projectionLayer, projectionLayer2] });
const projectionLayer2RTTProvider = layersFeature.createLayerWrapper(projectionLayer2).createRenderTargetProvider(sessionManager);
...
projectionLayer2RTTProvider.getRenderTargetForView(view);
```

I had to awkwardly store the generated layers with their types so that I could recover their types later in the session manager to create the correct corresponding wrapper (through the `WebXRLayerWrapperProvider` interface). 

The only solution I can think of is to have `updateRenderState` take a version of `XRRenderStateInit` that uses wrapped layers, but that would basically mean having a copy of the definition... I guess it's not as bad compared to all of the weird logic I have to support the current approach.